jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -286,33 +289,45 @@ protected boolean responseHeaders(HttpExchange exchange)
             return false;
 
         HttpResponse response = exchange.getResponse();
+        HttpFields responseHeaders = response.getHeaders();
         if (LOG.isDebugEnabled())
-            LOG.debug("Response headers {}{}{}", response, System.lineSeparator(), response.getHeaders().toString().trim());
-        ResponseNotifier notifier = getHttpDestination().getResponseNotifier();
-        List<Response.ResponseListener> responseListeners = exchange.getConversation().getResponseListeners();
-        notifier.notifyHeaders(responseListeners, response);
-        contentListeners.reset(responseListeners);
-        contentListeners.notifyBeforeContent(response);
+            LOG.debug("Response headers {}{}{}", response, System.lineSeparator(), responseHeaders.toString().trim());
 
-        if (!contentListeners.isEmpty())
+        // HEAD responses may have Content-Encoding
+        // and Content-Length, but have no content.
+        if (!HttpMethod.HEAD.is(exchange.getRequest().getMethod()))
         {
-            List<String> contentEncodings = response.getHeaders().getCSV(HttpHeader.CONTENT_ENCODING.asString(), false);
-            if (contentEncodings != null && !contentEncodings.isEmpty())
+            // Content-Encoding may have multiple values in the order they
+            // are applied, but we only support one decoding pass, the last one.
+            String contentEncoding = responseHeaders.get(HttpHeader.CONTENT_ENCODING);
+            if (contentEncoding != null)
             {
+                int comma = contentEncoding.indexOf(",");
+                if (comma > 0)
+                {
+                    QuotedCSV parser = new QuotedCSV(false);
+                    parser.addValue(contentEncoding);
+                    List<String> values = parser.getValues();
+                    contentEncoding = values.get(values.size() - 1);
+                }
```

Comment:
If we wanted to be really nasty, we should strip "identity" encodings off the end.... but the server doesn't do that (anymore)

Reply:
Most of the times `Content-Encoding` is single-valued. 

As you seem concerned about hot paths, I chose not to use `getCSV()` because in the single-value case allocates a `List` and a `QuotedCSV` for nothing, while the code above only does so for multi-values.

Doing a first `get()` and then `getCSV()` iterates over the headers twice, although should be a small list.

What do you prefer?

Reply:
It is a vexed question and now I'm concerned the server doesn't well handle the case of multiple `HttpField`s.

If we are struggling to come up with something that is both simple, clear and efficient, then probably the `HttpFields` API is not good enough.
Perhaps we need a method to ask `HttpFields` if it has a value as the last entry in a list of values, with the option of removing that entry.  This could be implemented efficiently without having to tokenize the list twice.    I'd be happy with an implementation that worked well for single valued lists, but was more expensive for multi valued lists.   

Something like:
```java
    String contentEncoding = responseHeaders.getLastCSV(HttpHeader.CONTENT_ENCODING);
```

and on mutable:
```java
    String contentEncoding = responseHeaders.takeLastCSV(HttpHeader.CONTENT_ENCODING);
```

Reply:
@sbordet I've created #10340 that might be useful hear as an efficient way to find the last value.    I've not done anything about removing the last value until I get feedback.

Reply:
@sbordet I've further update #10340 with a listIterator that make operating on the last item simpler.   So now in the server GzipRequest, I just copy the `HttpFields` to a mutable, then iterate in reverse over it correcting the headers found.
This will have merge hell with #10339 and #10308